flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/ChangelogStateHandleStreamImpl.java

```diff
@@ -83,11 +80,7 @@ public KeyedStateHandle getIntersection(KeyGroupRange keyGroupRange) {
 
     @Override
     public void discardState() throws Exception {
-        // discard only on TM; on JM, shared state is removed on subsumption
-        if (stateRegistry == null) {
-            bestEffortDiscardAllStateObjects(
-                    handlesAndOffsets.stream().map(t2 -> t2.f0).collect(Collectors.toList()));
-        }
+        // do nothing: rely on SharedStateRegistry for cleanup after ACK;
```

Comment:
When would the `handlesAndOffsets` be deleted then?

Reply:
The in-memory object will be GCed on checkpoint subsumption;
The actual handles will be discarded by `SharedStateRegistry` when no longer used (they are registered in `registerSharedStates`).

Reply:
Thanks for your clarification, I think we can improve the comment, e.g What the `ACK` is meaning here?

Reply:
I meant acknowledging checkpoint by task to JobManager.
I expaneded the comment:
```
// Do nothing: state will be discarded by SharedStateRegistry once JM receives it and a
// newer checkpoint completes without using it. JM might not receive the handle in the
// following cases:
// 1. hard TM failure
// 2. job termination
// 3. materialization of changes written pre-emptively
// The above cases will be addressed by FLINK-23139 and/or FLINK-24852``